Reset allowed usage and warn when maybeCreating configurations (#26467)
gradle/gradle: subprojects/docs/src/docs/userguide/jvm/building_java_projects.adoc

 * <<#sec:custom_java_source_set_paths,Customizing file and directory locations>>
 * <<java_testing.adoc#sec:configuring_java_integration_tests,Configuring Java integration tests>>

[[sec:implicit_sourceset_configurations]]
=== Source set configurations

Every source set that exists will cause Gradle to create several associated configurations as described above.
These configurations should **not** be created or accessed before Gradle does this.

> Gradle expects these associated configurations to allow only certain types of usage and will ideally create them with the proper usage.
> 
> When Gradle detects that configurations with these implicitly "reserved" names already exist, it will warn about this situation with a deprecation message.
> Gradle will then attempt to mutate the allowed usage to match the expected usage.

Reviewer: ```suggestion
When creating a source set, if one of these automatically created configurations already exists, Gradle will emit a deprecation warning. 
If the existing configuration's role is different than the role that the source set would have assigned, its role will be mutated to the correct value and another deprecation warning will be emitted. 
```

The build below demonstrates this unwanted behavior.

.Configurations created prior to their associated source sets
====
include::sample[dir="snippets/java/sourceSets-badConfNames/kotlin",files="build.gradle.kts[tags=java-dependency-mgmt]"]
include::sample[dir="snippets/java/sourceSets-badConfNames/groovy",files="build.gradle[tags=java-dependency-mgmt]"]
====

> This situation will be announced with the following deprecation warning:

Reviewer: ```suggestion
In this case, the following deprecation warning is emitted:
```

```
When creating configurations during sourceSet custom setup, Gradle found that configuration customCompileClasspath already exists with permitted usage(s):
	Consumable - this configuration can be selected by another project as a dependency
	Resolvable - this configuration can be resolved by this project to a set of files
	Declarable - this configuration can have dependencies added to it
Yet Gradle expected to create it with the usage(s):
	Resolvable - this configuration can be resolved by this project to a set of files
```

Following two simple best practices will avoid this problem:

> 1. Don't create configurations with names that will be used by source sets, such as names ending in `Api`, `Implementation`, `ApiElements`, `CompileOnly`, `CompileOnlyApi`, `RuntimeOnly`, `RuntimeClasspath` or `RuntimeElements`.

Reviewer: I think we should update the wording here to more than just `create`. The configuration container automatically creates configurations when they are accessed. For example if a user writes:

```
configurations {
    myCustomConfiguration {
        extendsFrom(implementation)
    }
}
```

The `implementation` configuration is created. The user did not intend to create this configuration. We should change the wording to `access or create`, or similar, ensure users do not ignore these guidelines

Author: Hmmm, IDK.  I think readers should understand from the tutorials, and from just using Gradle in general, that they are creating a new configuration in your example - especially if this is the first lexical appearance of `myCustomConfiguration` in their buildscript.

"Access" suggest to me the opposite of creation, when you're doing `configurations.getByName()` or something to find something that already exists - this is what we _want_ them to do.  

I think this change would actually confuse the guidance.

@lkasso - WDYT?

Reviewer 2: When you read this section, the context is around "creation" of the configuration.
We should leave item 1 as-is.

Author: Adding example.

(This list is not exhaustive.)

> 1. Create any custom source sets prior to any custom configurations.

Reviewer: There are two `1.`s in this list

Author: This is a markdown thing - it will always render 1, 2, 3, but you can just write it 1, 1, 1 so if you insert into the middle of the list you don't have to do tedious re-numbering.

include::sample[dir="snippets/java/sourceSets-badConfNames/groovy",files="build.gradle[tags=create-configurations]"]
====

In this case, the following deprecation warning is emitted:

```
When creating configurations during sourceSet custom setup, Gradle found that configuration customCompileClasspath already exists with permitted usage(s):
	Consumable - this configuration can be selected by another project as a dependency
	Resolvable - this configuration can be resolved by this project to a set of files
	Declarable - this configuration can have dependencies added to it
Yet Gradle expected to create it with the usage(s):
	Resolvable - this configuration can be resolved by this project to a set of files
```

Following two simple best practices will avoid this problem:

1. Don't create configurations with names that will be used by source sets, such as names ending in `Api`, `Implementation`, `ApiElements`, `CompileOnly`, `CompileOnlyApi`, `RuntimeOnly`, `RuntimeClasspath` or `RuntimeElements`.
(This list is not exhaustive.)
1. Create any custom source sets prior to any custom configurations.

> Remember that any time you reference a configuration within the `configurations` container - with or without supplying an initialization action - Gradle will create the configuration.

Author: Discussed.  Removing Kotlin example and clarifying Groovy.
